src/dmd/utils.d

```diff
@@ -138,3 +139,44 @@ const(char)[] toDString (const(char)* s) pure nothrow @nogc
 {
     return s ? s[0 .. strlen(s)] : null;
 }
+
+/**
+   Copy the content of `src` into a C-string ('\0' terminated) then call `dg`
+
+   The intent of this function is to provide an allocation-less
+   way to call a C function using a D slice.
+   The function internally allocates a buffer if needed, but frees it on exit.
+
+   Note:
+     The argument to `dg` is `scope`. To keep the data around after `dg` exits,
+     one has to copy it.
+
+   Params:
+     src = Slice to use to call the C function
+     dg  = Delegate to call afterwards
+
+   Returns:
+     The return value of `T`
+*/
+auto toCStringThen(alias dg)(const(char)[] src)
+{
+    char[512] small = void;
+    scope ptr = (src.length < (small.length - 1))
+                    ? small[0 .. src.length + 1]
+                    : (cast(char*)mem.xmalloc(src.length + 1))[0 .. src.length + 1];
+    scope (exit)
```

Comment:
Using scope in dmd is problematic - take a look at the code generated for it, it isn't pretty. It's a lot better if the function is marked nothrow.